### PR TITLE
Add automatically started per-database Maintenance Worker

### LINKED ISSUE #1421: DROP EXTENSION citus fails in SQL_DROP trigger
Set up some dummy SQL_DROP trigger that with a statement that goes through the planner hook:
```
CREATE TABLE public.unrelated_table (x int, y int);

CREATE OR REPLACE FUNCTION cleanup()
 RETURNS event_trigger
 LANGUAGE plpgsql
AS $function$                         
BEGIN
    /* go into citus planner hook */
   DELETE FROM public.unrelated_table;
END;                      
$function$;        
CREATE EVENT TRIGGER cleanup_trigger ON SQL_DROP
EXECUTE PROCEDURE cleanup();
```

Then in new session, `DROP EXTENSION citus` attempts to access `pg_dist_partition` (OID 51219) after it was dropped:
```
DROP EXTENSION citus; 
ERROR:  could not open relation with OID 51219
CONTEXT:  SQL statement "DELETE FROM public.unrelated_table"
PL/pgSQL function cleanup() line 3 at SQL statement
```

However, it succeeds the second time:
```
# DROP EXTENSION citus;
DROP EXTENSION
#
```

### PR DESCRIPTION
**Pull request by anarazel**
This'll be used for deadlock detection, 2pc recovery and other similar tasks.

Note that the two first commits fix various issues with cache maintenance.  If needed I can split them out into a separate PR.

See also #1394 /  #1447.

Fixes #1421

### CODE REVIEW

**Comment by lithp** [comment]
Neat, the cache changes are nice to finally have!

Why did you add a worker-per-database rather than have a global worker?

**Comment by marcocitus** [comment]
> Why did you add a worker-per-database rather than have a global worker?

A background worker can only interact with one database.

File: src/backend/distributed/utils/metadata_cache.c [hunk 1]
```diff
@@ -2221,6 +2253,20 @@ GetLocalGroupId(void)
 
 
 /*
+ * RegisterLocalGroupIdCacheCallbacks registers the callbacks required for the
+ * worker node cache.  It's separate from InitializeWorkerNodeCache so the
+ * callback can be registered early, before the tables exist.
```
Comment (marcocitus) [inline src/backend/distributed/utils/metadata_cache.c]: Comment is copied from `RegisterWorkerNodeCacheCallbacks` and needs to be adapted.

File: src/include/distributed/shared_library_init.h [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#ifndef SHARED_LIBRARY_INIT_H
+#define SHARED_LIBRARY_INIT_H
```
Comment (marcocitus) [inline src/include/distributed/shared_library_init.h]: I wish we were using `#pragma once`

File: src/backend/distributed/utils/maintenanced.c [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+	{
+		/* FIXME: better message, reference relevant guc in hint */
+		ereport(ERROR, (errmsg("ran out of database slots")));
+	}
+
+	if (!found || !dbData->daemonStarted)
+	{
+		BackgroundWorker worker;
+		BackgroundWorkerHandle *handle;
+		int pid;
```
Comment (marcocitus) [inline src/backend/distributed/utils/maintenanced.c]: Should have `= NULL` and `= 0` per convention

File: src/include/distributed/maintenanced.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/*-------------------------------------------------------------------------
+ *
+ * maintenanced.h
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
```
Comment (marcocitus) [inline src/include/distributed/maintenanced.h]: Should also be in `shared_library_init.h`, and maybe have some meaningful comment.

File: src/backend/distributed/utils/maintenanced.c [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+	HASHCTL hashInfo;
+	int hashFlags = 0;
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+
+	MaintenanceDaemonControl =
+		(MaintenanceDaemonControlData *) ShmemInitStruct("Citus Maintenance Daemon",
+														 MaintenanceDaemonShmemSize(),
+														 &alreadyInitialized);
+	if (!alreadyInitialized)
```
Comment (marcocitus) [inline src/backend/distributed/utils/maintenanced.c]: When is this already initialised?

**Comment by marcocitus** [comment]
Currently a `pg_ctl restart` gives the following log message:
```
LOG:  worker process: Citus Maintenance Daemon: 13322 (PID 1547) exited with exit code 1
```

**Comment by anarazel** [comment]
> Currently a pg_ctl restart gives the following log message:
> LOG:  worker process: Citus Maintenance Daemon: 13322 (PID 1547) exited with exit code 1

Right - that's to be expected?

File: src/include/distributed/maintenanced.h (hunk 4 above)
Comment (anarazel) [inline src/include/distributed/maintenanced.h]: Not sure what you mean here? Remove maintenanced.h?

File: src/backend/distributed/utils/maintenanced.c (hunk 5 above)
Comment (anarazel) [inline src/backend/distributed/utils/maintenanced.c]: On EXEC_BACKEND style builds (primarily windows), and crash restarts.

**Comment by marcocitus** [comment]
> Right - that's to be expected?

Why not exit cleanly?

File: src/backend/distributed/utils/maintenanced.c (hunk 5 above)
Comment (marcocitus) [inline src/backend/distributed/utils/maintenanced.c]: Might be worth adding a comment

File: src/include/distributed/maintenanced.h (hunk 4 above)
Comment (marcocitus) [inline src/include/distributed/maintenanced.h]: sorry, put 2 comments in 1.

1: maintenance.d should have a meaningful comment at the top
2: shared_library_init.h should have a copyright notice.

**Comment by anarazel** [comment]
> Why not exit cleanly?

Two reasons: exit(1) is the "traditional" response to SIGTERM in all other processes. Secondly, exit(0) unregisters a bgworker, i.e. it'll not be restarted anymore, which is not what we'd want to do as a general response to SIGTERM.

**Comment by marcocitus** [comment]
> exit(1) is the "traditional" response to SIGTERM in all other processes. 

Except postgres background processes, which all exit(0).

> Secondly, exit(0) unregisters a bgworker, Secondly, exit(0) unregisters a bgworker, i.e. it'll not be restarted anymore, which is not what we'd want to do as a general response to SIGTERM.

Why not stop on SIGTERM and allow restart via `InitializeMaintenanceDaemonBackend`?

**Comment by anarazel** [comment]
> Except postgres background processes, which all exit(0).

Well there postmaster has special logic to handle exits of such processes, so that's not applicable here.  The applicable comparison is to bgworkers and plain backends.

> >  Secondly, exit(0) unregisters a bgworker, i.e. it'll not be restarted anymore, which is not what we'd want to do as a general response to SIGTERM.

> Why not stop on SIGTERM and allow restart via InitializeMaintenanceDaemonBackend?

Because then it'd not be started unless new connections are established?  And even then there'd be a race condition around the maintenanced quitting and new connections being established.   Just seems to add complications for no gain.

**Comment by marcocitus** [comment]
> Because then it'd not be started unless new connections are established? 

Which is the case anyway.

I'm not concerned about convention here. It's just a bad and confusing user experience as it's indistinguishable from a crash.

**Comment by anarazel** [comment]
>  > Because then it'd not be started unless new connections are established?
> Which is the case anyway.

Huh? No, it gets restarted automatically after 5 secs.


> I'm not concerned about convention here. It's just a bad and confusing user experience as it's indistinguishable from a crash.

exit(1) is never a crash?

**Comment by marcocitus** [comment]
> Huh? No, it gets restarted automatically after 5 secs.

When postgres starts then "it'd not be started unless new connections are established".

> exit(1) is never a crash?

I'm calling an error that causes the background worker to have to restart a crash here.

**Comment by anarazel** [comment]
I'm really not folllowing.  Right now maintenanced gets restarted both after restarts due to manual SIGTERM and due to ERROR/FATALs it raises internally.  That seems to be the right behaviour for something kind of critical functionality-wise.   You seem to be advocating that we change either or both of these to not-restart and instead rely on maintenanced being restarted when there's a new connection to a citus using database?

**Comment by marcocitus** [comment]
This also fixes #1421

File: src/backend/distributed/utils/metadata_cache.c [hunk 6]
```diff
@@ -81,27 +82,39 @@ typedef struct ShardCacheEntry
...
+	Oid distShardPlacementNodeidIndexId;
+	Oid distTransactionRelationId;
+	Oid distTransactionGroupIndexId;
+	Oid extraDataContainerFuncId;
+	Oid workerHashFunctionId;
+	Oid extensionOwner;
+};
+
+
+static struct CachedData CachedData;
```
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: Done (as of next push).

File: src/backend/distributed/utils/metadata_cache.c (hunk 1 above)
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: Done (as of next push).

File: src/include/distributed/shared_library_init.h (hunk 2 above)
Comment (anarazel) [inline src/include/distributed/shared_library_init.h]: That's a holy war I'm not going to get involved into on either side ;)

File: src/backend/distributed/utils/maintenanced.c (hunk 3 above)
Comment (anarazel) [inline src/backend/distributed/utils/maintenanced.c]: Done (as of next push).

File: src/backend/distributed/utils/maintenanced.c (hunk 5 above)
Comment (anarazel) [inline src/backend/distributed/utils/maintenanced.c]: Done (as of next push).

File: src/include/distributed/maintenanced.h (hunk 4 above)
Comment (anarazel) [inline src/include/distributed/maintenanced.h]: Done (as of next push).

**Comment by anarazel** [comment]
> This also fixes #1421

Heh, nice accident.

**Comment by anarazel** [comment]
I've updated the patch to address the points you had, including adding an error context.  It's right now just saying
```
	errcontext("Citus maintenance daemon for database %u user %u",
			   myDbData->databaseOid, myDbData->userOid);
```

File: src/backend/distributed/utils/maintenanced.c [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+		BackgroundWorkerHandle *handle = NULL;
+		int pid = 0;
+
+		dbData->userOid = extensionOwner;
+
+		snprintf(worker.bgw_name, BGW_MAXLEN,
+				 "Citus Maintenance Daemon: %u/%u",
+				 MyDatabaseId, extensionOwner);
+
+		memset(&worker, 0, sizeof(worker));
```
Comment (marcocitus) [inline src/backend/distributed/utils/maintenanced.c]: This needs to be above `snprintf(worker.bgw_name, ...`.

**Comment by marcocitus** [comment]
I'm just going to push a fix for that `multi_reference_table` test that keeps failing on travis to master.

File: src/backend/distributed/utils/maintenanced.c (hunk 7 above)
Comment (anarazel) [inline src/backend/distributed/utils/maintenanced.c]: That, uh, is a very fair complaint.  WTH?

**Comment by anarazel** [comment]
Thanks for the review!

**Comment by anarazel** [comment]
Thanks for the review!

File: src/include/distributed/shared_library_init.h (hunk 2 above)
Comment (jasonmp85) [inline src/include/distributed/shared_library_init.h]: Is there anyone on the other side?

**Pull request merged**